src/FactorGraphTypes.jl

```diff
@@ -132,18 +132,24 @@ $(TYPEDEF)
 
 TODO remove Union types -- issue #383
 """
-mutable struct FactorMetadata
-  factoruserdata
-  variableuserdata::Union{Vector, Tuple}
-  variablesmalldata::Union{Vector, Tuple}
-  solvefor::Union{Symbol, Nothing}
-  variablelist::Union{Nothing, Vector{Symbol}}
-  dbg::Bool
-  FactorMetadata() = new() # [], []
-  FactorMetadata(x1, x2::Union{Vector,Tuple},x3) = new(x1, x2, x3, nothing, nothing, false)
-  FactorMetadata(x1, x2::Union{Vector,Tuple},x3,x4::Symbol) = new(x1, x2, x3, x4, nothing, false)
-  FactorMetadata(x1, x2::Union{Vector,Tuple},x3,x4::Symbol,x5::Vector{Symbol};dbg::Bool=false) = new(x1, x2, x3, x4, x5, dbg)
+mutable struct FactorMetadata{T}
+  factoruserdata # deprecate, not in use
+  variableuserdata::Union{Vector, Tuple} # deprecate
+  variablesmalldata::Union{Vector, Tuple} # deprecate
+  solvefor::Union{Symbol, Nothing} # Change to Symbol?
+  variablelist::Union{Nothing, Vector{Symbol}} # Vector{Symbol} #deprecate
+  dbg::Bool #
+    cachedata::Union{Nothing,Vector{T}} # New. change to Vector{T}
```

Comment:
might be simpler to use only `::Vector{T}` and drop the Union?  This does work:
```julia
julia> Vector{Nothing}
Array{Nothing,1}
```

This should give better type stability?

Reply:
https://github.com/JuliaRobotics/IncrementalInference.jl/issues/783#issuecomment-665080114

Reply:
I want to keep Union Nothing around for now. It allows nothing to be used in the vector for other cases. Say you do not want to cache something.